### PR TITLE
CE 555

### PR DESCRIPTION
**Pull request by serishema**
CE 555

* Tested crashdumps by adding a null pointer dereference to bats so it would crash as soon as there was an alive bat in the level. It was correctly identified as the cause of the crash in the dump.
* rotating the log files on crash ended up being enough.
* I wasn't able to test this on Linux.

Sorry, clang-format has made a bit of a mess of the diff for croftengine.cpp - I hope you can follow what I actually changed.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,13 @@ list( APPEND CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake )
 
 project( CroftEngine VERSION 1.9.1 )
 
+if (WIN32) 
+    set(CMAKE_CONFIGURATION_TYPES Debug RelWithDebInfo CACHE TYPE INTERNAL FORCE )
```
Comment (stohrendorf) [inline CMakeLists.txt:7]: What's this line supposed to do?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -706,6 +706,12 @@ add_subdirectory( core )
 add_subdirectory( launcher )
 add_subdirectory( dosbox-cdrom )
 
+if (WIN32) 
+        set( WIN32_SPECIFIC_LIBS dbghelp)
```
Comment (stohrendorf) [inline src/CMakeLists.txt:710]: That won't work, as it's part of the Windows SDK, which naturally isn't installed on common PCs.

File: src/croftengine.cpp [hunk 3]
```diff
@@ -31,27 +32,12 @@
...
-  BOOST_LOG_TRIVIAL(fatal) << "Abnormal termination";
-  stacktrace::logStacktrace();
-
-  if(oldTerminateHandler != nullptr)
-    oldTerminateHandler();
-}
 
 const gsl::czstring logFormat = "[%TimeStamp% %Severity% %ThreadID%] %Message%";
 
+typedef boost::log::sinks::synchronous_sink<boost::log::sinks::text_file_backend> sink_t;
```
Comment (stohrendorf) [inline src/croftengine.cpp:38]:
```suggestion
using sink_t = boost::log::sinks::synchronous_sink<boost::log::sinks::text_file_backend>;
```

File: src/croftengine.cpp [hunk 4]
```diff
@@ -69,20 +55,61 @@ void initConsole(boost::log::trivial::severity_level level)
...
                            boost::log::keywords::format = logFormat,
                            boost::log::keywords::auto_flush = true,
                            boost::log::keywords::max_files = 10);
 }
+
+bool initCrashReporting()
+{
+  auto& chillout = Debug::Chillout::getInstance();
+
+  if(const auto userDataDir = findUserDataDir(); userDataDir.has_value())
```
Comment (stohrendorf) [inline src/croftengine.cpp:70]: Please invert the condition so it becomes an early exit.

File: src/croftengine.cpp [hunk 5]
```diff
@@ -69,20 +55,61 @@ void initConsole(boost::log::trivial::severity_level level)
...
+    }
+#ifdef WIN32
+    chillout.init(L"croftengine", crashdumpDir);
+#elif
+    chillout.init("croftengine", crashdumpDir);
+#endif
+    chillout.setBacktraceCallback(
+      [](const char* stackTraceEntry)
+      {
+        BOOST_LOG_TRIVIAL(error) << stackTraceEntry;
```
Comment (stohrendorf) [inline src/croftengine.cpp:85]:
```suggestion
        BOOST_LOG_TRIVIAL(fatal) << stackTraceEntry;
```

File: src/croftengine.cpp [hunk 6]
```diff
@@ -69,20 +55,61 @@ void initConsole(boost::log::trivial::severity_level level)
...
+      });
+    chillout.setCrashCallback(
+      [&chillout]()
+      {
+        BOOST_LOG_TRIVIAL(error) << "Croft engine has crashed, writing minidump";
```
Comment (stohrendorf) [inline src/croftengine.cpp:90]:
```suggestion
        BOOST_LOG_TRIVIAL(fatal) << "Croft engine has crashed, writing minidump";
```

**Review by stohrendorf — CHANGES_REQUESTED** [review]
There are more locations where `logStackTrace` is called (`utils.cpp`, `gslfailhandler.cpp`, ``). These can be removed I think.

File: CMakeLists.txt (hunk 1 above)
Comment (serishema) [reply CMakeLists.txt:7]: It prevents building Release with a multi-configuration build system like Visual Studio.

File: src/CMakeLists.txt (hunk 2 above)
Comment (serishema) [reply src/CMakeLists.txt:710]: That used to be the case, but all currently supported versions of windows include dbghelp.dll (see https://docs.microsoft.com/en-us/windows/win32/debug/dbghelp-versions). 
To make sure I tried this on a PC without visual studio or the windows SDK installed and it worked.

**Comment by serishema** [comment]
I've deleted the old logStackTrace to make sure there are no calls to it left in.

File: src/CMakeLists.txt (hunk 2 above)
Comment (stohrendorf) [reply src/CMakeLists.txt:710]: Ah, didn't know that, I only remember having issues with it a few years ago.

**Review by stohrendorf — APPROVED** [review]
Looks OK as far as I can see. Didn't test it yet, though.

**Pull request merged**